dhall-json/src/Dhall/JSONToDhall.hs

```diff
@@ -369,13 +380,257 @@ typeCheckSchemaExpr compileException expr =
+
+{-| `inferSchema` will never infer a union type with more than one numeric
+    alternative
+
+    Instead, the most general alternative type will be preferred, which this
+    type tracks
+-}
+data UnionNumber
+    = UnionAbsent
+    -- ^ The union type does not have a numeric alternative
+    | UnionNatural
+    -- ^ The union type has a @Natural@ alternative
+    | UnionInteger
+    -- ^ The union type has an @Integer@ alternative
+    | UnionDouble
+    -- ^ The union type has a @Double@ alternative
+    deriving (Bounded, Eq, Ord)
+
+-- | Unify two numeric alternative types by preferring the most general type
+instance Semigroup UnionNumber where
+    (<>) = max
+
+instance Monoid UnionNumber where
+    mempty = minBound
+
+unionNumberToAlternatives :: UnionNumber -> [ (Text, Maybe (Expr s a)) ]
+unionNumberToAlternatives UnionAbsent  = []
+unionNumberToAlternatives UnionNatural = [ ("Natural", Just D.Natural) ]
+unionNumberToAlternatives UnionInteger = [ ("Integer", Just D.Integer) ]
+unionNumberToAlternatives UnionDouble  = [ ("Double" , Just D.Double ) ]
+
+{-| A union type that `inferSchema` can infer
+
+    This type will have at most three alternatives:
+
+    * A @Bool@ alternative
+    * Either a @Natural@, @Integer@, or @Double@ alternative
+    * A @Text@ alternative
+
+    These alternatives will always use the same names and types when we convert
+    back to a Dhall type, so we only need to keep track of whether or not each
+    alternative is present.
+-}
+data UnionSchema = UnionSchema
+    { bool :: Any
+    -- ^ `True` if the union has a @Bool@ alternative
+    , number :: UnionNumber
+    -- ^ Up to one numeric alternative
+    , text :: Any
+    -- ^ `True` if the union has a @Text@ alternative
+    } deriving (Eq)
+
+unionSchemaToDhallType :: UnionSchema -> Expr s a
+unionSchemaToDhallType UnionSchema{..} = D.Union (Map.fromList alternatives)
+  where
+    alternatives =
+            (if getAny bool then [ ("Bool", Just D.Bool) ] else [])
+        <>  unionNumberToAlternatives number
+        <>  (if getAny text then [ ("Text", Just D.Text) ] else [])
+
+-- | Unify two union types by combining their alternatives
+instance Semigroup UnionSchema where
+    UnionSchema boolL numberL textL <> UnionSchema boolR numberR textR =
+        UnionSchema{..}
+      where
+        bool = boolL <> boolR
+
+        number = numberL <> numberR
+
+        text = textL <> textR
+
+instance Monoid UnionSchema where
+    mempty = UnionSchema{..}
+      where
+        bool = mempty
+
+        number = mempty
+
+        text = mempty
+
+{-| A `Schema` is a subset of the `Expr` type representing all possible
+    Dhall types that `inferSchema` could potentially return
+-}
+data Schema
+    = Bool
+    | Natural
+    | Integer
+    | Double
+    | Text
+    | List Schema
+    | Optional Schema
+    | Record RecordSchema
+    | Union UnionSchema
+    | ArbitraryJSON
+
+-- | (`<>`) unifies two schemas
+instance Semigroup Schema where
```

Comment:
I wonder whether there's a more concise way to define a commutative `(<>)` like here…

Reply:
@sjakobi: There is a way to do it.  I considered adding a fall-through case of `x <> y = y <> x` to halve the number of alternatives, but then decided against it, because I wasn't sure that I wouldn't accidentally introduce an infinite loop.